### ISSUE TITLE
dynamically adjust batch_size or provide batch_size in bytes in nlp.pipe

### ISSUE THREAD

**Issue by AlJohri** [explosion/spaCy#4858]
## Feature description

The goal is optimally use resources when using `nlp.pipe` in a multiprocessing setting. Currently, in order to adjust `batch_size` it requires multiple calls to `nlp.pipe` with manually partitioned documents. This has the overhead of forking processes or re-loading models in each process (not sure how its currently implemented) for every invocation of nlp.pipe.

I would like a way to dynamically adjust the `batch_size` based on the document length. **If I could provide the `batch_size` as bytes this would achieve the same affect.**

I'm currently having a lot of trouble with this as I need to manually adjust both batch_size AND n_processes given the diverse document lengths in my corpus. Some documents require using 1/3 or 1/4 of the maximum concurrency at a batch_size of 1 while others can use max concurrency with a `batch_size` of 15-20. It would be great to have a way to more easily handle this.

P.S. I am using `neuralcoref` in case that impacts the memory consumption calculations.

## Could the feature be a [custom component](https://spacy.io/usage/processing-pipelines#custom-components) or [spaCy plugin](https://spacy.io/universe)?

I don't think so?

**Comment by svlandeg** [comment]
I'm trying to think how that feature could potentially be implemented without disrupting the current code base & existing scripts too much. It feels (to me) like this should be some sort of pre-processing step, which I guess you have already implemented (I'm guessing you're not actually doing it "manually" ;-)).

Another thing that we recently ran into ourselves is that memory requirements may also depend on the length of sentences in a doc. By some unlucky turn of events, you may end up with a random batch of data all containing very long sentences that can't get processed, while a different batch with the same number of instances, or the same length in terms of bytes, has no problem.

So in short I'm not sure if there's a universal answer to this. Maybe it depends on the use-case how the data should be organized and batched up.

But if you do have a suggestion on how this could potentially be implemented, perhaps you can sketch that out a bit?
